ProcessLib/TwoPhaseFlowWithPrho/TwoPhaseFlowWithPrhoLocalAssembler.h

```diff
@@ -0,0 +1,200 @@
+    // copy the previous to current
+    void pushBackState()
+    {
+        _sw_pre = _sw;
+        _rho_m_pre = _rho_m;
+    }
+};
+namespace ProcessLib
+{
+namespace TwoPhaseFlowWithPrho
+{
+const unsigned NUM_NODAL_DOF = 2;
+
+class TwoPhaseFlowWithPrhoLocalAssemblerInterface
+    : public ProcessLib::LocalAssemblerInterface,
+      public NumLib::ExtrapolatableElement
+{
+public:
+    virtual std::vector<double> const& getIntPtSaturation(
+        std::vector<double>& /*cache*/) const = 0;
+
+    virtual std::vector<double> const& getIntPtNonWettingPressure(
+        std::vector<double>& /*cache*/) const = 0;
+};
+
+template <typename ShapeFunction, typename IntegrationMethod,
+          unsigned GlobalDim>
+class TwoPhaseFlowWithPrhoLocalAssembler
+    : public TwoPhaseFlowWithPrhoLocalAssemblerInterface
+{
+    using ShapeMatricesType = ShapeMatrixPolicyType<ShapeFunction, GlobalDim>;
+    using ShapeMatrices = typename ShapeMatricesType::ShapeMatrices;
+
+    using LocalAssemblerTraits = ProcessLib::LocalAssemblerTraits<
+        ShapeMatricesType, ShapeFunction::NPOINTS, NUM_NODAL_DOF, GlobalDim>;
+
+    using NodalMatrixType = typename ShapeMatricesType::NodalMatrixType;
+    using NodalVectorType = typename ShapeMatricesType::NodalVectorType;
+    using GlobalDimMatrixType = typename ShapeMatricesType::GlobalDimMatrixType;
+    using GlobalDimVectorType = typename ShapeMatricesType::GlobalDimVectorType;
+    using LocalMatrixType = typename LocalAssemblerTraits::LocalMatrix;
+    using LocalVectorType = typename LocalAssemblerTraits::LocalVector;
+
+public:
+    TwoPhaseFlowWithPrhoLocalAssembler(
+        MeshLib::Element const& element,
+        std::size_t const /*local_matrix_size*/,
+        bool const is_axially_symmetric,
+        unsigned const integration_order,
+        TwoPhaseFlowWithPrhoProcessData const& process_data)
+        : _element(element),
+          _integration_method(integration_order),
+          _shape_matrices(initShapeMatrices<ShapeFunction, ShapeMatricesType,
+                                            IntegrationMethod, GlobalDim>(
+              element, is_axially_symmetric, _integration_method)),
+          _process_data(process_data),
+          _saturation(
+              std::vector<double>(_integration_method.getNumberOfPoints())),
+          _pressure_nonwetting(
+              std::vector<double>(_integration_method.getNumberOfPoints()))
+    {
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+        _ip_data.reserve(n_integration_points);
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            _ip_data.emplace_back(*_process_data._material);
+            _ip_data[ip]._sw = 1.0;
+            _ip_data[ip]._rho_m = 0.0;
+            _ip_data[ip]._dsw_dpg = 0.0;
+            _ip_data[ip]._dsw_drho = 0.0;
+            _ip_data[ip]._drhom_dpg = 0.0;
+            _ip_data[ip]._drhom_drho = 0.0;
```

Comment:
These initializations belong into the IntegrationPointData constructor's initialization list.